File: flame-crypto.js
import "crypto";
import { readFile } from "fs/promises";

/**
 * @param {ArrayBuffer} buf 
 */
const arrayBufferToBase64 = (buf) => {
    return btoa(String.fromCharCode.apply(null, new Uint8Array(buf)));
}

const readPublicKey = async (path) => {
    const fileContents = (await readFile(path)).toString("ascii")
        .replace("-----BEGIN PUBLIC KEY-----", "")
        .replace("-----END PUBLIC KEY-----", "");
    
    const buf = Buffer.from(fileContents, "base64");

    return await crypto.subtle.importKey(
        "spki",
        buf,
        {
            "name": "ECDH",
            "namedCurve": "P-384",
        },
        true,
        []
    );
}

const readPrivateKey = async (path) => {
    const fileContents = (await readFile(path)).toString("ascii")
        .replace("-----BEGIN PRIVATE KEY-----", "")
        .replace("-----END PRIVATE KEY-----", "");
    
    const buf = Buffer.from(fileContents, "base64");

    return await crypto.subtle.importKey(
        "pkcs8",
        buf,
        {
            "name": "ECDH",
            "namedCurve": "P-384",
        },
        true,
        ["deriveKey", "deriveBits"],
    );
}

/**
 * @param {CryptoKey} pubkey 
 * @param {CryptoKey} privkey 
 */
const deriveKey = async (pubkey, privkey) => {
    return await crypto.subtle.deriveKey(
        {
            name: "ECDH",
            public: pubkey
        },
        privkey,
        {
            name: "AES-GCM",
            length: 256
        },
        true,
        ["encrypt", "decrypt"]
    );
}

const deriveBits = async (pubkey, privkey) => {
    return await crypto.subtle.deriveBits(
        {
            name: "ECDH",
            public: pubkey
        },
        privkey,
        384
    );
}

const theirPubKey = await readPublicKey("alice.pem");
const myPrivKey = await readPrivateKey("bob.pfx");
const secretKey = await deriveKey(theirPubKey, myPrivKey);

const sharedSecretKey = await arrayBufferToBase64(await crypto.subtle.exportKey("raw", secretKey));
const sharedBits = arrayBufferToBase64(await deriveBits(theirPubKey, myPrivKey));

const myPubKey = await readPublicKey("bob.pem");
const theirPrivKey = await readPrivateKey("alice.pfx");

const otherSecretKey = await deriveKey(theirPubKey, myPrivKey);

console.log("sharedSecretKey", sharedSecretKey);
console.log("sharedBits", sharedBits);

console.log("sk match", sharedSecretKey == await arrayBufferToBase64(await crypto.subtle.exportKey("raw", otherSecretKey)));
console.log("bits match", sharedBits == arrayBufferToBase64(await deriveBits(theirPubKey, myPrivKey)));

const iv = new Uint8Array(96 / 8);
crypto.getRandomValues(iv);

const message = "Hello world!";

const encryptedMessage = await crypto.subtle.encrypt(
    {
        name: "AES-GCM",
        iv: iv
    },
    secretKey,
    Buffer.from(message, "ascii")
);

console.log("encrypted message", encryptedMessage);

const decryptedMessage = await crypto.subtle.decrypt(
    {
        name: "AES-GCM",
        iv: iv
    },
    otherSecretKey,
    encryptedMessage
);

console.log("decrypted message", new TextDecoder("ascii").decode(decryptedMessage));

File: flame-crypto.py
from cryptography.hazmat.primitives import serialization
from cryptography.hazmat.primitives.asymmetric import ec
from cryptography.hazmat.primitives.ciphers.aead import AESGCM

import base64, os

def main():
    with open("alice.pem", mode="rb") as f:
        their_pubkey = serialization.load_pem_public_key(f.read())
    
    with open("bob.pem", mode="rb") as f:
        my_pubkey = serialization.load_pem_public_key(f.read())

    with open("alice.pfx", mode="rb") as f:
        their_privkey = serialization.load_pem_private_key(f.read(), password=None)

    with open("bob.pfx", mode="rb") as f:
        my_privkey = serialization.load_pem_private_key(f.read(), password=None)

    shared_secret = my_privkey.exchange(ec.ECDH(), their_pubkey)
    other_shared_secret = their_privkey.exchange(ec.ECDH(), my_pubkey)

    sk_txt = base64.b64encode(shared_secret[:32]).decode("utf-8")
    sb_txt = base64.b64encode(shared_secret).decode("utf-8")

    print("sharedSecretKey", sk_txt)
    print("sharedBits", sb_txt)

    print("sk match", sk_txt == base64.b64encode(other_shared_secret[:32]).decode("utf-8"))
    print("bits match", sb_txt == base64.b64encode(other_shared_secret).decode("utf-8"))

    # shared secret has 384 bits. to use it with aes gcm, key must be truncated to 256 bits.
    aesgcm = AESGCM(shared_secret[:32])
    other_aesgcm = AESGCM(other_shared_secret[:32])

    iv = os.urandom(96 // 8)
    message = "Hello world!"
    encrypted_message = aesgcm.encrypt(iv, message.encode("ascii"), b"")
    print("encrypted message", encrypted_message)

    decrypted_message = other_aesgcm.decrypt(iv, encrypted_message, b"")
    print("decrypted message", decrypted_message)


if __name__ == "__main__":
    main()
